File: webmvp/src/lib/sessionNavigation.test.ts
import { describe, expect, it } from "vitest";

import {
  buildAdjacentSongHref,
  parseSessionNavParams,
  sessionSongHref,
  startSetHref,
} from "@/lib/sessionNavigation";
import type { SessionSong } from "@/lib/types";

const songs: SessionSong[] = [
  {
    id: "a",
    songId: "song-1",
    songTitle: "Song One",
    order: 0,
    keyOverride: "G",
    notes: null,
    addedBy: "u1",
    addedAt: {} as SessionSong["addedAt"],
  },
  {
    id: "b",
    songId: "song-2",
    songTitle: "Song Two",
    order: 1,
    keyOverride: null,
    notes: null,
    addedBy: "u1",
    addedAt: {} as SessionSong["addedAt"],
  },
];

describe("sessionNavigation", () => {
  it("parses playlist params", () => {
    const params = new URLSearchParams("playlist=s1&index=1&key=G");
    expect(parseSessionNavParams(params)).toEqual({
      sessionId: "s1",
      index: 1,
    });
  });

  it("parses legacy session params", () => {
    const params = new URLSearchParams("session=s1&index=1&key=G");
    expect(parseSessionNavParams(params)).toEqual({
      sessionId: "s1",
      index: 1,
    });
  });

  it("builds song href with playlist context", () => {
    expect(sessionSongHref("s1", songs[0], 0)).toBe(
      "/song/song-1?playlist=s1&index=0&key=G",
    );
  });

  it("builds start set href", () => {
    expect(startSetHref("s1", songs)).toBe(
      "/song/song-1?playlist=s1&index=0&key=G",
    );
  });

  it("builds adjacent hrefs", () => {
    expect(buildAdjacentSongHref("s1", songs, 1, -1)).toBe(
      "/song/song-1?playlist=s1&index=0&key=G",
    );
    expect(buildAdjacentSongHref("s1", songs, 0, 1)).toBe(
      "/song/song-2?playlist=s1&index=1",
    );
    expect(buildAdjacentSongHref("s1", songs, 0, -1)).toBeNull();
  });
});

File: mobile/test/domain/session_navigation_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:lf_chords/domain/session_navigation.dart';

void main() {
  final songs = [
    SessionSongEntry(
      id: 'a',
      songId: 'song-1',
      songTitle: 'Song One',
      order: 0,
      keyOverride: 'G',
    ),
    SessionSongEntry(
      id: 'b',
      songId: 'song-2',
      songTitle: 'Song Two',
      order: 1,
    ),
  ];

  test('parses playlist params', () {
    final params = parseSessionNavParams({'playlist': 's1', 'index': '1', 'key': 'G'});
    expect(params.sessionId, 's1');
    expect(params.index, 1);
  });

  test('parses legacy session params', () {
    final params = parseSessionNavParams({'session': 'legacy', 'index': '0'});
    expect(params.sessionId, 'legacy');
    expect(params.index, 0);
  });

  test('canonicalPlaylistQuery rewrites session to playlist', () {
    expect(
      canonicalPlaylistQuery({'session': 's1', 'index': '2'}),
      {'index': '2', 'playlist': 's1'},
    );
    expect(canonicalPlaylistQuery({'playlist': 's1'}), isNull);
  });

  test('SessionNavParams value equality', () {
    const a = SessionNavParams(sessionId: 's1', index: 2);
    const b = SessionNavParams(sessionId: 's1', index: 2);
    const c = SessionNavParams(sessionId: 's1', index: 3);
    expect(a, b);
    expect(a == c, isFalse);
    expect(a.hashCode, b.hashCode);
  });

  test('builds song path with playlist context', () {
    expect(
      sessionSongPath('s1', songs[0], 0),
      '/song/song-1?playlist=s1&index=0&key=G',
    );
  });

  test('builds adjacent paths', () {
    expect(
      buildAdjacentSongPath('s1', songs, 1, -1),
      '/song/song-1?playlist=s1&index=0&key=G',
    );
    expect(buildAdjacentSongPath('s1', songs, 0, -1), isNull);
  });
}
